src/zarr/common.py

```diff
@@ -137,17 +137,21 @@ def parse_named_configuration(
 
 def parse_shapelike(data: int | Iterable[int]) -> tuple[int, ...]:
     if isinstance(data, int):
+        if data < 0:
+            raise ValueError(f"Expected a non-negative integer. Got {data} instead")
         return (data,)
-    if not isinstance(data, Iterable):
-        raise TypeError(f"Expected an iterable. Got {data} instead.")
-    data_tuple = tuple(data)
-    if len(data_tuple) == 0:
-        raise ValueError("Expected at least one element. Got 0.")
+    try:
+        data_tuple = tuple(data)
```

Comment:
we should also be checking that all values are integers here. NumPy raises an error if you pass a float as a shape

```
In [4]: np.zeros((4., ))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[4], line 1
----> 1 np.zeros((4., ))

TypeError: 'float' object cannot be interpreted as an integer
```

Reply:
This is addressed by [32ba006](https://github.com/zarr-developers/zarr-python/pull/1979/commits/32ba006334eeb7a746978d65af18d222f0ff9d21) and [e866f24](https://github.com/zarr-developers/zarr-python/pull/1979/commits/e866f247877c5cea462709e3d8fa1f8741ef59e0)